src/core/net/netif.hpp

```diff
@@ -152,8 +153,16 @@ class NetifMulticastAddress : public otNetifMulticastAddress,
         return static_cast<NetifMulticastAddress *>(const_cast<otNetifMulticastAddress *>(mNext));
     }
 
+#if (OPENTHREAD_FTD || OPENTHREAD_MTD) && OPENTHREAD_CONFIG_MLR_ENABLE
+    MlrState GetMlrState() const { return mMlrState; }
+    void     SetMlrState(MlrState aState) { mMlrState = aState; }
+#endif
 private:
     bool Matches(const Address &aAddress) const { return GetAddress() == aAddress; }
+
+#if (OPENTHREAD_FTD || OPENTHREAD_MTD) && OPENTHREAD_CONFIG_MLR_ENABLE
+    MlrState mMlrState : 2;
+#endif
```

Comment:
We are adding new member variable in `NetifMulticastAddress` class which is inheriting from `otNetifMulticastAddress`

We should be careful that we are not getting an `otNetifMulticastAddress` object (not really being `NetifMulticastAddress`) and then down-casting/treating it as `NetifMulticastAddress`.  I think this pattern is not safe. 

I know we do  have some fixed/constant multicast addresses (like "link-local all nodes", "link-local all routers", etc.) which are defined as constant `otNetifMulticastAddress` objects (and stored in flash) - defined in `netif.cpp` source file.

I undestand that the way `mMleState` is being used later in the code (we do check if it is external address), it may work ok but I think using this pattern may set a bad precedence. 

So I feel we may be better not add `mMleState` to the  `NetifMulticastAddress` class.

Thoughts?

Reply:
Thinking about this I can suggest a possible different approach:
- We keep `NetifMultciastAddress` as is.
- We define a sub-class of it which then adds the new member variable and getter/setter for this (maybe calling it `ExternalNetifMultciastAddress` or `MlrNetifMultciastAddress`?)
- We can then add a new method to `Netif` to iterate through all external/MLR multicast addresses (only external/MLR ones) and return a pointer/reference to new sub-class type.
- User of this method can then safely get/set `MleState`.

Thoughts?

Reply:
> I know we do have some fixed/constant multicast addresses (like "link-local all nodes", "link-local all routers", etc.) which are defined as constant `otNetifMulticastAddress` objects (and stored in flash) - defined in `netif.cpp` source file.

Implementing a link list with some elements being constant is challenging. Do we make sure that the constant `otNetifMulticastAddress`s are always at the tail of the list? @abtink

Reply:
> * We define a sub-class of it which then adds the new member variable and getter/setter for this (maybe calling it `ExternalNetifMultciastAddress` or `MlrNetifMultciastAddress`?)

I actually thought of it too. It should even save us some memory (MlrState on constant `otNetifMulticastAddress` is redundant).

Reply:
> Implementing a link list with some elements being constant is challenging. Do we make sure that the constant otNetifMulticastAddresss are always at the tail of the list? @abtink 
 
Indeed. They are always at the tail of the list. This whole model is used mainly to reduce RAM use.

https://github.com/openthread/openthread/blob/ceb63f97d57a42f15359c2fb1a6b912c2a963221/src/core/net/netif.cpp#L45-L55
You can check the implementation of `SubscribeAllNodesMulticast()` and `SubscribeAllRoutersMulticast()`, etc for more details (there are comments to help make the code easier to read and follow).

Reply:
Another thought is that we may define one MlrState array with the size same as `mExtMulticastAddressPool`, per MA that requires MLR.req, we can get the index, and use the index to sync the MlrState correspondingly.

Comparatively, this option might bring less code changes on the multicast address class between different Thread versions, though  the subclass option is more C++ style.

Reply:
This can work as well but we need to define the `MlrState` array in the `Netif` itself, since the fact that we use an array/pool for external addresses and therefore can have an index associated with each address  is internal to `Netif` class (not part of `public` definitions of `Netif`).

Thinking more about this, I feel having a way to iterate over the external (only) addresses would be useful in general. We can use this in some place in the code (e.g., in `Mle::AppendAddressRegistration()` where we go through the list of multicast address but do skip over entries that are not external).

- We can define `ExtNetifMulticastAddress` as sub-class of `NetifMulticastAddress`.
- We can overload the `GetNext()` method in the sub-class `ExtNetifMulticastAddress` to get the next external address entry (basically have `GetNext()` implementation skip over linked-list entries that are not external).
- This would be simple for user of `Netif`, they can the head of a linked-list of all `NetifMulticastAddress` entries or the head of linked-list of `ExtNetifMulticastAddress` and in both cases, they iterate through it like a linked-list.

Thoughts?

Reply:
My opinion on the two solutions: 

> 1# Add `ExternalNetifMulticastAddress` as a new subclass of `NetifMulticastAddress` with `MlrState` being it's member. 
> 2# Add an of MlrState with equal size of  mExtMulticastAddressPool.

First of all, I think they are similar solutions without much practical difference. 
- 1# is more object oriented by encapsulating MlrState in the `ExternalNetifMulticastAddress` class, but it requires significant code refactoring. For example,  `ExternalNetifMulticastAddress` will be put to a `LinkedList<ExternalNetifMulticastAddress>` but `ExternalNetifMulticastAddress.GetNext()` returns a `NetifMulticastAddress`, which violates the protocol of `LinkedList`. 
- 2# introduces much less change to existing code but it assumes that a unique index can be get for each external multicast address (e.g. allocated from a array of a Pool). 2# also allows using a bit-vector with 2 bits for each address to further reduce the memory consumption, even though the benefit is minor (e.g. 8 multicast addresses consumes 8B in solution 1# but only 2B in solution 2#). 

Overall I prefer 2# for a little bit because it involves less code change. 

Thoughts? @abtink @librasungirl @jwhui

Reply:
> * We can overload the `GetNext()` method in the sub-class `ExtNetifMulticastAddress` to get the next external address entry (basically have `GetNext()` implementation skip over linked-list entries that are not external).

Overloading should be necessary, but skipping over non-external entries seems unnecessary. `ExtNetifMulticastAddress.GetNext()` is only used for the free list of the pool, so when it's called, the next entry **MUST** be an ExtNetifMulticastAddress entry or nullptr.

Reply:
@simonlingoogle , all options are fine for me. 

> It requires significant code refactoring. For example, `ExternalNetifMulticastAddress` will be put to a `LinkedList<ExternalNetifMulticastAddress>` but `ExternalNetifMulticastAddress.GetNext()` returns a NetifMulticastAddress, which violates the protocol of LinkedList.

What I had in mind  was not to create a new actual linked-list of `LinkedList<ExternalNetifMulticastAddress>`. I agree that would not be good and would require a lot of changes.

The idea was that we really have one actual linked-list in memory which is a list of all `NetifMulticastAddress` objects. We define a new sub-class `ExtNetifMulticastAdderss` where we just overload the `GetNext()` to go through the same linked-list but skip over all entries that are not external (i.e., make it look like as if the external addresses are in a linked-list of their own).

 But considering how to implement this, I realized it may not be as easy as I initially thought - main thing is that we cannot tell in `(Ext)NetifMulticastAdderss` class itself whether an entry is external or not. For this, we need to have have access to `Netif` class - which we can get if we make `ExtNetifMulticastAddress` be an `InstanceLocatorInit` (all doable  but requires more changes).

Reply:
I pushed in the last commit an implementation of `ExternalNetifMulticastAddress` class (actually I have got this code finished before @librasungirl suggested the alternative solution). @abtink @librasungirl 

It allows us to use for-loops like this to iterate all external multicast addresses:
```c++
    for (Ip6::ExternalNetifMulticastAddress &addr : Get<ThreadNetif>().IterateExternalMulticastAddresses()) { ... }
```
I think this code can be useful for other purposes too.